packages/components/src/dropdown-menu/index.tsx

```diff
@@ -62,13 +139,18 @@ function DropdownMenu( dropdownMenuProps ) {
 	}
 
 	// Normalize controls to nested array of objects (sets of controls)
-	let controlSets;
+	let controlSets: NonNullable< typeof controls >[];
 	if ( controls?.length ) {
+		// @ts-expect-error The check below is needed because `DropdownMenus`
+		// rendered by `ToolBarGroup` receive controls as a nested array.
```

Comment:
As far as I can tell, `ToolbarGroup` rendered `DropdownMenus` only pass controls as nested arrays because they [already undergo the exact same normalization process](https://github.com/WordPress/gutenberg/blob/d1f5b8646364ae1384256b51523670c2d86455a9/packages/components/src/toolbar/toolbar-group/index.js#L76-L80). I'm not sure if that normalization is actually needed or not, but it might be worth looking into when `ToolbarGroup` is worked on. Pinging @margolisj, as I know you [mentioned](https://github.com/WordPress/gutenberg/pull/49190#issuecomment-1527678557) doing some work there.

Reply:
Thanks for the ping. #50262 branch has the code I've been working on. Very early work though not sold on any of it.
These components have been a bit and abused with various overrides instead of new components. In terms of that normalization code, I'm not sure about the functionality as I've mostly worked around it. I did make a small change to the following if though. Also, I considered adding DropdownMenu to the ToolbarGroup work iirc, so perhaps with both of these being related we could change the code as opposed to just typing it?

Reply:
> In terms of that normalization code, I'm not sure about the functionality as I've mostly worked around it.

I did a little more digging to try and understand the functionality of that normalization. Basically it looks like `ToolbarGroup` [is set up to apply a divider between control sets when it receives a nested array](https://github.com/WordPress/gutenberg/blob/e948f63cc44a7d85fb581e2ab8765c512fe37f2e/packages/components/src/toolbar/toolbar-group/index.js#L103-L106) of controls. There's even [a unit test that specifically confirms that behavior is working properly](https://github.com/WordPress/gutenberg/blob/e948f63cc44a7d85fb581e2ab8765c512fe37f2e/packages/components/src/toolbar/test/toolbar-group.tsx#L61).

I threw together a quick local test in Storybook and it isn't _pretty_, but it does work. [You can see the relevant styles here](https://github.com/WordPress/gutenberg/blob/39dbf19e5844983302ace6ea4c88f35ab639c923/packages/components/src/toolbar/toolbar-group/style.scss#L67-L85).

<img width="255" alt="A screenshot of an example ToolbarGroup. It shows a 'bold' format button and an 'italic' format button. In between the two a light grey separator is displayed, but it is very faint and not vertically centered." src="https://user-images.githubusercontent.com/13856531/236280080-0c9dde90-34dd-4d97-bd67-9bcb2dbb27bb.png">

Since this does appear to have some functional purpose (no idea if/how often it's actually used, but that's a separate question) I feel like we should probably leave it as-is for now? If anything, we could address it in a followup if we were to decide that we wanted to either simplify and remove, or update the styling of those separators so they look a bit nicer. It looks like it'd be enough of a runtime change that it wouldn't fit well in your TypeScript refactor, so I'd say keep working around it for now... thanks for brainstorming with me!

Reply:
> we could change the code as opposed to just typing it

Our preference for TypeScript refactors is to avoid introducing runtime changes, 

> I feel like we should probably leave it as-is for now? If anything, we could address it in a followup if we were to decide that we wanted to either simplify and remove, or update the styling of those separators so they look a bit nicer.

This would be the preferred approach!